File: python/wn.py
import random
import re
from collections import defaultdict


class markov:
    def __init__(self):
        self.chain = defaultdict(list)
        self.start_words = []

    def clean_text(self, text):
        text = text.replace("\n", " ")
        text = re.sub(r'\s+', ' ', text)
        return text

    def train(self, text):
        text = self.clean_text(text)
        words = text.split()

        for i in range(len(words) - 2):
            w1, w2, w3 = words[i], words[i+1], words[i+2]
            self.chain[(w1, w2)].append(w3)

            if w1.endswith(".") and w2[0].isupper():
                self.start_words.append((w2, w3))

    def generate(self, length=80):
        if self.start_words and random.random() > 0.2:
            w1, w2 = random.choice(self.start_words)
        else:
            w1, w2 = random.choice(list(self.chain.keys()))

        result = [w1, w2]

        for _ in range(length):
            key = (w1, w2)

            if key not in self.chain:
                break

            choices = self.chain[key]

            if len(choices) > 1:
                next_word = random.choice(choices + random.sample(choices, len(choices)))
            else:
                next_word = choices[0]

            result.append(next_word)

            if next_word.endswith("."):
                break

            w1, w2 = w2, next_word

        return " ".join(result)


with open("wn.txt", "r", encoding="utf-8") as f:
    text = f.read()

ai = markov()
ai.train(text)

print(ai.generate(80))

File: rust/wn.rs
use rand::prelude::*;
use std::collections::HashMap;
use std::fs;
use std::sync::Arc;

type Word = Arc<str>;
type ChainKey = (Word, Word);

pub struct Markov {
    chain: HashMap<ChainKey, Vec<Word>>,
    start_words: Vec<ChainKey>,
}

impl Markov {
    pub fn new() -> Self {
        Self {
            chain: HashMap::new(),
            start_words: Vec::new(),
        }
    }

    fn clean_text(text: &str) -> String {
        text.replace('\n', " ")
            .split_whitespace()
            .collect::<Vec<_>>()
            .join(" ")
    }

    pub fn train(&mut self, raw_text: &str) {
        let cleaned = Self::clean_text(raw_text);
        let words: Vec<Word> = cleaned.split_whitespace().map(Into::into).collect();

        if words.len() < 3 {
            return;
        }

        for i in 0..words.len() - 2 {
            let w1 = Arc::clone(&words[i]);
            let w2 = Arc::clone(&words[i + 1]);
            let w3 = Arc::clone(&words[i + 2]);

            let key = (w1, w2);

            if key.0.ends_with('.') && key.1.chars().next().map_or(false, |c| c.is_uppercase()) {
                self.start_words.push((Arc::clone(&key.1), Arc::clone(&w3)));
            }

            self.chain.entry(key).or_insert_with(Vec::new).push(w3);
        }
    }

    pub fn generate(&self, length: usize) -> String {
        let mut rng = rand::rng();

        let mut current_key = if !self.start_words.is_empty() && rng.random_bool(0.8) {
            self.start_words.choose(&mut rng).unwrap().clone()
        } else {
            let keys: Vec<&ChainKey> = self.chain.keys().collect();
            (*keys.choose(&mut rng).expect("Chain is empty")).clone()
        };

        let mut result = vec![current_key.0.to_string(), current_key.1.to_string()];

        for _ in 0..length {
            if let Some(choices) = self.chain.get(&current_key) {
                let next_word = choices.choose(&mut rng).unwrap();
                result.push(next_word.to_string());

                if next_word.ends_with('.') {
                    break;
                }

                current_key = (Arc::clone(&current_key.1), Arc::clone(next_word));
            } else {
                break;
            }
        }

        result.join(" ")
    }
}

fn main() {
    let content = fs::read_to_string("wn.txt").expect("Could not read file");
    let mut ai = Markov::new();
    ai.train(&content);
    println!("{}", ai.generate(80));
}
